File: ticket_management_system/resources/flights.py
"""Flight management API endpoints."""
from flask import Blueprint, request, jsonify
from marshmallow import ValidationError
from ticket_management_system.extensions import cache
from ticket_management_system.utils import handle_validation_error, handle_general_error, handle_conflict_error
from ticket_management_system.resources.users import token_required, admin_required
from ticket_management_system.static.schema.flight_schemas import FlightSearchSchema, AddFlightSchema, UpdateFlightSchema
from ticket_management_system.resources.flight_service import FlightService
from ticket_management_system.exceptions import FlightAlreadyExistsError, FlightNotFoundError

flight_bp = Blueprint('flights', __name__, url_prefix='/api/flights')


@flight_bp.route('/airports', methods=['GET'])
@token_required
@cache.cached(timeout=50)
def get_airports(_current_user):
    """Get list of available airports."""
    try:
        result = FlightService.get_available_airports()
        return jsonify(result), 200
    except Exception as e:  # pylint: disable=broad-exception-caught
        return handle_general_error(e, rollback=False)


@flight_bp.route('/search', methods=['GET'])
@token_required
def search_flights(_current_user):
    """Search flights with filters."""
    try:
        # Validate query parameters using Marshmallow schema
        schema = FlightSearchSchema()
        validated_data = schema.load(request.args)

        # Extract validated parameters
        status = validated_data.get('status')
        origin_airport = validated_data.get('origin_airport')
        destination_airport = validated_data.get('destination_airport')
        departure_date = validated_data.get('departure_date')
        arrival_date = validated_data.get('arrival_date')
        page = validated_data.get('page', 1)
        per_page = validated_data.get('per_page', 10)
        sort_by = validated_data.get('sort_by', 'departure_time')
        sort_order = validated_data.get('sort_order', 'asc')

        # Convert date objects to strings for service layer
        departure_date_str = departure_date.strftime('%Y-%m-%d') if departure_date else None
        arrival_date_str = arrival_date.strftime('%Y-%m-%d') if arrival_date else None

        # Search flights using service
        result = FlightService.search_flights(
            status=status,
            origin_airport=origin_airport,
            destination_airport=destination_airport,
            departure_date=departure_date_str,
            arrival_date=arrival_date_str,
            page=page,
            per_page=per_page,
            sort_by=sort_by,
            sort_order=sort_order,
        )

        return jsonify(result), 200
    except ValidationError as err:
        return handle_validation_error(err)
    except Exception as e:  # pylint: disable=broad-exception-caught
        return handle_general_error(e, rollback=False)


@flight_bp.route('/', methods=['POST'])
@token_required
@admin_required
def add_flight(_current_user):
    """Add a new flight (admin only)."""
    try:
        # Get and validate request body using Marshmallow schema
        schema = AddFlightSchema()
        validated_data = schema.load(request.get_json())

        # Create flight using service
        new_flight = FlightService.create_flight(
            flight_code=validated_data['flight_code'],
            origin_airport=validated_data['origin_airport'],
            destination_airport=validated_data['destination_airport'],
            departure_time=validated_data['departure_time'],
            arrival_time=validated_data['arrival_time'],
            base_price=validated_data['base_price']
        )

        # Format response
        flight_data = FlightService.format_flight_detail(new_flight)
        response = {
            'flight': flight_data,
            'message': 'Flight created successfully'
        }

        return jsonify(response), 201
    except ValidationError as err:
        return handle_validation_error(err)
    except FlightAlreadyExistsError as err:
        return handle_conflict_error(err.message)
    except Exception as e:  # pylint: disable=broad-exception-caught
        return handle_general_error(e)


@flight_bp.route('/<uuid:flight_id>', methods=['GET'])
@token_required
def get_flight(_current_user, flight_id):
    """Get a single flight by ID."""
    try:
        flight = FlightService.get_flight_by_id(flight_id)
        if not flight:
            raise FlightNotFoundError(flight_id)
        return jsonify({'flight': FlightService.format_flight_detail(flight)}), 200
    except FlightNotFoundError as err:
        return jsonify({
            'error': 'Not Found',
            'message': err.message
        }), 404
    except Exception as e:  # pylint: disable=broad-exception-caught
        return handle_general_error(e, rollback=False)


@flight_bp.route('/<uuid:flight_id>', methods=['DELETE'])
@token_required
@admin_required
def delete_flight(_current_user, flight_id):
    """Delete a flight by ID."""
    try:
        # Delete flight using service
        FlightService.delete_flight(flight_id)
        return jsonify({
            'message': f'Flight {flight_id} deleted successfully'
        }), 200
    except FlightNotFoundError as err:
        return jsonify({
            'error': 'Not Found',
            'message': err.message
        }), 404
    except Exception as e:  # pylint: disable=broad-exception-caught
        return handle_general_error(e)


@flight_bp.route('/<uuid:flight_id>', methods=['PUT'])
@token_required
@admin_required
def update_flight(_current_user, flight_id):
    """Update a flight by ID (admin only)."""
    try:
        # Get and validate request body using Marshmallow schema
        schema = UpdateFlightSchema()
        validated_data = schema.load(request.get_json())

        # Update flight using service
        updated_flight = FlightService.update_flight(
            flight_id,
            **validated_data
        )

        # Format response
        flight_data = FlightService.format_flight_detail(updated_flight)
        response = {
            'flight': flight_data,
            'message': 'Flight updated successfully'
        }

        return jsonify(response), 200
    except ValidationError as err:
        return handle_validation_error(err)
    except FlightNotFoundError as err:
        return jsonify({
            'error': 'Not Found',
            'message': err.message
        }), 404
    except Exception as e:  # pylint: disable=broad-exception-caught
        return handle_general_error(e)

File: flight-client-app/src/services/flights.ts
import { apiClient } from "../lib/apiClient";
import type { Flight, PaginatedResponse } from "../types";

export type FlightSearchFilters = {
  status?: "all" | "active" | "inactive" | "started" | "en_route" | "landed" | "cancelled" | "delayed";
  origin_airport?: string;
  destination_airport?: string;
  departure_date?: string;
  arrival_date?: string;
  page?: number;
  per_page?: number;
  sort_by?: "departure_time" | "arrival_time" | "base_price";
  sort_order?: "asc" | "desc";
};

function sanitizeFlightSearchFilters(filters: FlightSearchFilters): Record<string, string | number> {
  return Object.fromEntries(
    Object.entries(filters).filter(([, value]) => value !== "" && value != null),
  ) as Record<string, string | number>;
}

export const searchFlights = async (filters: FlightSearchFilters): Promise<Flight[]> => {
  const response = await apiClient.get("/api/flights/search", {
    params: sanitizeFlightSearchFilters(filters),
  });
  const data = response.data as PaginatedResponse<Flight>;
  return data.flights ?? data.items ?? data.data ?? [];
};

export const searchFlightsPaginated = async (
  filters: FlightSearchFilters,
): Promise<{ flights: Flight[]; pagination: NonNullable<PaginatedResponse<Flight>["pagination"]> }> => {
  const response = await apiClient.get("/api/flights/search", {
    params: sanitizeFlightSearchFilters(filters),
  });
  const data = response.data as PaginatedResponse<Flight>;
  return {
    flights: data.flights ?? data.items ?? data.data ?? [],
    pagination: data.pagination ?? {
      page: filters.page ?? 1,
      per_page: filters.per_page ?? 10,
      total_pages: 1,
      total_items: 0,
      has_next: false,
      has_prev: false,
      next_page: null,
      prev_page: null,
    },
  };
};

export const searchAllFlights = async (filters: FlightSearchFilters): Promise<Flight[]> => {
  const perPage = 100;
  let page = 1;
  const allFlights: Flight[] = [];

  while (true) {
    const { flights, pagination } = await searchFlightsPaginated({
      ...filters,
      page,
      per_page: perPage,
    });
    allFlights.push(...flights);
    if (!pagination.has_next) break;
    page += 1;
  }

  return allFlights;
};

export const listAirports = async (): Promise<string[]> => {
  const response = await apiClient.get("/api/flights/airports");
  const data = response.data as { airports?: string[] };
  return data.airports ?? [];
};

type AdminFlightPayload = {
  flight_code: string;
  origin_airport: string;
  destination_airport: string;
  departure_time: string;
  arrival_time: string;
  base_price: number;
  status?: "active" | "inactive" | "started" | "en_route" | "landed" | "cancelled" | "delayed";
};

export const createFlight = async (payload: AdminFlightPayload): Promise<Flight> => {
  const response = await apiClient.post("/api/flights/", payload);
  return (response.data as { flight?: Flight }).flight as Flight;
};

export const updateFlight = async (
  flightId: string,
  payload: Partial<AdminFlightPayload>,
): Promise<Flight> => {
  const response = await apiClient.put(`/api/flights/${flightId}`, payload);
  return (response.data as { flight?: Flight }).flight as Flight;
};

export const deleteFlight = async (flightId: string): Promise<void> => {
  await apiClient.delete(`/api/flights/${flightId}`);
};

export async function getFlightById(flightId: string): Promise<Flight | null> {
  try {
    const response = await apiClient.get(`/api/flights/${flightId}`);
    const data = response.data as { flight?: Flight };
    return data.flight ?? null;
  } catch {
    return null;
  }
}

export async function getFlightsByIds(flightIds: string[]): Promise<Map<string, Flight>> {
  const unique = [...new Set(flightIds.filter(Boolean))];
  const map = new Map<string, Flight>();
  const results = await Promise.allSettled(unique.map((id) => getFlightById(id)));
  for (let i = 0; i < unique.length; i++) {
    const result = results[i];
    if (result.status === "fulfilled" && result.value) {
      map.set(unique[i], result.value);
    }
  }
  return map;
}

